File: Python/031 Next Permutation.py
# Implement next permutation, which rearranges numbers into the lexicographically next greater permutation of numbers.

# If such arrangement is not possible, it must rearrange it as the lowest possible order (ie, sorted in ascending order).

# The replacement must be in-place and use only constant extra memory.

# Here are some examples. Inputs are in the left-hand column and its corresponding outputs are in the right-hand column.

# 1,2,3 → 1,3,2
# 3,2,1 → 1,2,3
# 1,1,5 → 1,5,1


class Solution(object):
    def nextPermutation(self, nums):
        """
        :type nums: List[int]
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        length = len(nums)
        i = length - 2
        while i >= 0 and nums[i] >= nums[i + 1]:
            i -= 1
        if i >= 0:
            j = i + 1
            while j <= length:
                if j == length or nums[i] >= nums[j]:
                    j -= 1
                    nums[i], nums[j] = nums[j], nums[i]
                    break
                j += 1
        i += 1
        j = length - 1
        while i < j:
            nums[i], nums[j] = nums[j], nums[i]
            i += 1
            j -= 1

# test
s = Solution()
arr = [1, 2, 3]
s.nextPermutation(arr)
print(arr)

arr = [3, 2, 1]
s.nextPermutation(arr)
print(arr)

arr = [1, 1, 5]
s.nextPermutation(arr)
print(arr)

arr = [1, 3, 2]
s.nextPermutation(arr)
print(arr)

File: JavaScript/031 Next Permutation.js
// Implement next permutation, which rearranges numbers into the lexicographically next greater permutation of numbers.

// If such arrangement is not possible, it must rearrange it as the lowest possible order (ie, sorted in ascending order).

// The replacement must be in-place and use only constant extra memory.

// Here are some examples. Inputs are in the left-hand column and its corresponding outputs are in the right-hand column.

// 1,2,3 → 1,3,2
// 3,2,1 → 1,2,3
// 1,1,5 → 1,5,1


// 从后向前遍历数组,直到找到两个相邻升序排列的元素(nums[i] < nums[i + 1])
// 然后向后查找第一个大于nums[i]的元素nums[j], 交换nums[i]和nums[j]
// 最后翻转[i + 1, len - 1]中的元素
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */

var nextPermutation = function (nums) {
    let swap = function (nums, i, j) {
        temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }

    let len = nums.length;
    let i = len - 2;

    while (i >= 0 && nums[i] >= nums[i + 1]) i--;
    if (i >= 0) {
        let j = i + 1;
        while (j <= len) {
            if (j === len || nums[i] >= nums[j]) {
                j--;
                swap(nums, i, j);
                break;
            }
            j++;
        }
    }
    i++;
    j = len - 1;
    while (i < j) {
        swap(nums, i, j);
        i++;
        j--;
    }
};

// test
let arr = [1, 2, 3];
nextPermutation(arr);
console.log(arr);

arr = [3, 2, 1];
nextPermutation(arr);
console.log(arr);

arr = [1, 1, 5];
nextPermutation(arr);
console.log(arr);

arr = [1, 3, 2];
nextPermutation(arr);
console.log(arr);